src/extract/str.rs

```diff
@@ -1,6 +1,13 @@
-use extract::{Extract, Error, Context, Immediate};
+use extract::{Context, Error, Extract, Immediate};
+use std::borrow::Cow;
+use url::percent_encoding;
 use util::BufStream;
 
+fn decode(s: &str) -> Result<String, Error> {
+    let percent_decoded = Cow::from(percent_encoding::percent_decode(s.as_bytes())).into_owned();
+    String::from_utf8(percent_decoded).map_err(|e| Error::invalid_argument(&e))
```

Comment:
It seems like you can use [`decode_utf8`](https://docs.rs/percent-encoding/1.0.1/percent_encoding/struct.PercentDecode.html#method.decode_utf8). Untested:

```rust
percent_encoding::percent_decode(s.as_bytes()))
    .decode_utf8()
    .map(Cow::into_owned)
    .map_err(...)
```

Reply:
Fixed, thanks.